drivers/bm/bm_dfu_core.c

```
@@ -0,0 +1,342 @@
+                }
+                break;
+            default:
+                break;
+        }
+    }
+}
+
+/**
+ * BM DFU Subsystem Thread (Runs HFSM in an event-driven manner)
+ */
+static void bm_dfu_subsystem_thread(void)
+{
+    int ret;
+
+    LOG_DBG("BM DFU Subsystem thread started");
+
+    while (1)
+    {
+        _dfu_context.current_event.type = DFU_EVENT_NONE;
+        k_msgq_get(&_dfu_subsystem_queue, &_dfu_context.current_event, K_FOREVER);
+
+        ret = smf_run_state(SMF_CTX(&_dfu_context));
+        if (ret) 
+        {
+            LOG_ERR("State Machine Error: terminating");
+            break;
+        }
+    }
+}
+
+static int bm_dfu_init( const struct device *arg )
+{
+    k_tid_t thread_id;
+    bm_dfu_event_t evt;
+    int retval;
+
+    ARG_UNUSED(arg);
+
+    /* Set initial state */
+    smf_set_initial(SMF_CTX(&_dfu_context), &dfu_states[BM_DFU_STATE_INIT]);
+
+    /* Create Transport Service thread to create events for HFSM */
+    thread_id = k_thread_create(&_transport_service_dfu_thread_data, _bm_dfu_transport_service_stack,
+            K_THREAD_STACK_SIZEOF(_bm_dfu_transport_service_stack),
+            (k_thread_entry_t)bm_dfu_transport_service_thread,
+            NULL, NULL, NULL, K_PRIO_COOP(10), 0, K_NO_WAIT);
+
+    if (thread_id == NULL)
+    {
+        LOG_ERR("DFU Transport Service thread not created");
+        return -1;
+    }
+
+    /* Thread to run HFSM */
+    thread_id = k_thread_create(&_subsystem_dfu_thread_data, _bm_dfu_subsystem_stack,
+            K_THREAD_STACK_SIZEOF(_bm_dfu_subsystem_stack),
+            (k_thread_entry_t)bm_dfu_subsystem_thread,
+            NULL, NULL, NULL, K_PRIO_COOP(10), 0, K_NO_WAIT);
+
+    if (thread_id)
+    {
+        evt.type = DFU_EVENT_INIT_SUCCESS;
+        retval = k_msgq_put(&_dfu_subsystem_queue, &evt, K_NO_WAIT);
+        if (retval)
+        {
+            LOG_ERR("Message could not be added to Queue");
```

Should we return an error here?